src/sentry/static/sentry/app/views/groupEvents.jsx

```diff
@@ -10,42 +11,90 @@ import Pagination from '../components/pagination';
 import SearchBar from '../components/searchBar';
 import EventsTable from '../components/eventsTable/eventsTable';
 import {t} from '../locale';
+import withEnvironment from '../utils/withEnvironment';
+import {getQueryEnvironment, getQueryStringWithEnvironment} from '../utils/queryString';
+import EnvironmentStore from '../stores/environmentStore';
+import {setActiveEnvironment} from '../actionCreators/environments';
 
 const GroupEvents = createReactClass({
   displayName: 'GroupEvents',
+
+  propTypes: {
+    environment: PropTypes.object,
+  },
+
   mixins: [ApiMixin, GroupState],
 
   getInitialState() {
-    let queryParams = this.props.location.query;
-    return {
+    const queryParams = this.props.location.query;
+
+    const initialState = {
       eventList: [],
       loading: true,
       error: false,
       pageLinks: '',
       query: queryParams.query || '',
     };
+
+    // If an environment is specified in the query, update the global environment
+    // Otherwise if a global environment is present update the query
+    const queryEnvironment = EnvironmentStore.getByName(
+      getQueryEnvironment(queryParams.query || '')
+    );
+
+    if (queryEnvironment) {
+      setActiveEnvironment(queryEnvironment);
+    } else if (this.props.environment) {
+      initialState.query = getQueryStringWithEnvironment(
+        initialState.query,
+        this.props.environment.name
+      );
+    }
+
+    return initialState;
   },
 
   componentWillMount() {
     this.fetchData();
   },
 
   componentWillReceiveProps(nextProps) {
-    if (
-      nextProps.params.groupId !== this.props.params.groupId ||
-      nextProps.location.search !== this.props.location.search
-    ) {
-      let queryParams = nextProps.location.query;
+    // If query has changed, update the environment with the query environment
```

Comment:
Is it intentional that this is no longer handling the case where `groupId` is changed?

Reply:
Yes, it's guarding against a situation that I can't see ever arising. We don't change group ID within the group details view. To change group ID you'd have to navigate to a different route and the component will be remounted.

Reply:
Seems I added it when upgrading to react-router 1.0 [here](https://github.com/getsentry/sentry/commit/b5571593ecb066567dcef3eaf42fac520626c7eb#diff-323aa0c7a08b4dc4c847b8897a152085R60). I don't really remember what I was doing; maybe I wasn't using routes correctly. I agree it should be re-mounted.